vtkext/private/module/F3DColoringInfoHandler.cxx

```diff
@@ -0,0 +1,187 @@
+{
+  // XXX: This assumes importer do not import actors with an empty input
+  assert(dataset);
+
+  // Recover all possible names
+  std::set<std::string> arrayNames;
+
+  vtkDataSetAttributes* attr = useCellData
+    ? static_cast<vtkDataSetAttributes*>(dataset->GetCellData())
+    : static_cast<vtkDataSetAttributes*>(dataset->GetPointData());
+
+  for (int i = 0; i < attr->GetNumberOfArrays(); i++)
+  {
+    vtkDataArray* array = attr->GetArray(i);
+    if (array && array->GetName())
+    {
+      arrayNames.insert(array->GetName());
+    }
+  }
+
+  auto& data = useCellData ? this->CellDataColoringInfo : this->PointDataColoringInfo;
+
+  for (const std::string& arrayName : arrayNames)
+  {
+    // Recover/Create a coloring info
+    F3DColoringInfoHandler::ColoringInfo& info = data[arrayName];
+    info.Name = arrayName;
+
+    vtkDataArray* array = useCellData ? dataset->GetCellData()->GetArray(arrayName.c_str())
+      : dataset->GetPointData()->GetArray(arrayName.c_str());
+    if (array)
+    {
+      info.MaximumNumberOfComponents =
+        std::max(info.MaximumNumberOfComponents, array->GetNumberOfComponents());
+
+      // Set ranges
+      // XXX this does not take animation into account
+      std::array<double, 2> range;
+      array->GetRange(range.data(), -1);
+      info.MagnitudeRange[0] = std::min(info.MagnitudeRange[0], range[0]);
+      info.MagnitudeRange[1] = std::max(info.MagnitudeRange[1], range[1]);
+
+      for (size_t i = 0; i < static_cast<size_t>(array->GetNumberOfComponents()); i++)
+      {
+        array->GetRange(range.data(), static_cast<int>(i));
+        if (i < info.ComponentRanges.size())
+        {
+          info.ComponentRanges[i][0] = std::min(info.ComponentRanges[i][0], range[0]);
+          info.ComponentRanges[i][1] = std::max(info.ComponentRanges[i][1], range[1]);
+        }
+        else
+        {
+          info.ComponentRanges.emplace_back(range);
+        }
+      }
+
+      // Set component names
+      if (array->HasAComponentName())
+      {
+        for (size_t i = 0; i < static_cast<size_t>(array->GetNumberOfComponents()); i++)
+        {
+          const char* compName = array->GetComponentName(i);
+          if (i < info.ComponentNames.size())
+          {
+            if (compName && info.ComponentNames[i] != std::string(compName))
+            {
+              // set non-coherent component names to empty string
+              info.ComponentNames[i] = "";
+            }
+          }
+          else
+          {
+            // Add components names to the back of the component names vector
+            info.ComponentNames.emplace_back(compName ? compName : "");
+          }
+        }
+      }
+    }
+  }
+}
+
+//----------------------------------------------------------------------------
+std::optional<F3DColoringInfoHandler::ColoringInfo> F3DColoringInfoHandler::SetCurrentColoring(bool enable, bool useCellData, std::optional<std::string> arrayName, bool quiet)
```

Comment:
That's a lot of `bool` arguments, which will generate cryptic calls. Can we split this function in several calls like `SetEnableColoring`, `SetColoringUseCellData`, `SetColoringArrayName` or does it have to be in the same function?

Also, `arrayName` should be const ref.

Reply:
It is possible but then I need to add a method that actually "triggers" the array lookup. That doesnt seems necessary to me as this method is very much internal to our vtkext private module and used only in vtkF3DRenderer.

Agreed on the const ref